fili-core/src/main/java/com/yahoo/bard/webservice/metadata/DataSourceMetadataLoader.java

```diff
@@ -175,20 +188,78 @@ protected void queryDataSourceMetadata(ConcretePhysicalTable table) {
-                } catch (IOException e) {
-                    throw new UnsupportedOperationException(
-                            DRUID_METADATA_READ_ERROR.format(table.getFactTableName()),
-                            e
-                    );
-                }
+        return buildDataSourceMetadataSuccessCallback(table.getDataSourceName());
+    }
+
+    /**
+     * Callback to parse druid datasource metadata response.
+     * <p>
+     * Typical druid datasource metadata response:
+     * <pre>
+     *  """
+     *  {
+     *      "name": "tableName",
+     *      "properties": { },
+     *      "segments": [
+     *          {
+     *              "dataSource": "tableName",
+     *              "interval": "2015-01-01T00:00:00.000Z/2015-01-02T00:00:00.000Z",
+     *              "version": "2015-01-15T18:08:20.435Z",
+     *              "loadSpec": {
+     *                  "type": "hdfs",
+     *                  "path": "hdfs:/some_hdfs_URL/tableName/.../index.zip"
+     *              },
+     *              "dimensions": "color", "shape",
+     *              "metrics": "height", "width",
+     *              "shardSpec": {
+     *                  "type":"hashed",
+     *                  "partitionNum": 0,
+     *                  "partitions": 2
+     *              },
+     *              "binaryVersion":9,
+     *              "size":1024,
+     *              "identifier":"tableName_2015-01-01T00:00:00.000Z_2015-01-02T00:00:00.000Z_2015-02-15T18:08:20.435Z"
+     *          },
+     *          {
+     *              "dataSource": "tableName",
+     *              "interval": "2015-01-01T00:00:00.000Z/2015-01-02T00:00:00.000Z",
+     *              "version": "2015-02-01T07:02:05.912Z",
+     *              "loadSpec": {
+     *                  "type": "hdfs",
+     *                  "path": "hdfs:/some_hdfs_URL/tableName/.../index.zip"
+     *              },
+     *              "dimensions": "color", "shape",
+     *              "metrics": "height", "width",
+     *              "shardSpec": {
+     *                  "type":"hashed",
+     *                  "partitionNum": 1,
+     *                  "partitions": 2
+     *              },
+     *              "binaryVersion":9,
+     *              "size":512,
+     *              "identifier":"tableName_2015-01-01T00:00:00.000Z_2015-01-02T00:00:00.000Z_2015-02-01T07:02:05.912Z"
+     *          }
+     *      ]
+     *   }"""
+     * </pre>
+     *
+     * @param dataSourceName  The datasource name to inject into this callback.
+     *
+     * @return The callback itself.
+     */
+    protected SuccessCallback buildDataSourceMetadataSuccessCallback(DataSourceName dataSourceName) {
+        return rootNode -> {
+            try {
+                DataSourceMetadata dataSourceMetadata = mapper.treeToValue(rootNode, DataSourceMetadata.class);
+                metadataService.update(dataSourceName, dataSourceMetadata);
```

Comment:
Would it be better to combine this line and it's previous line into 1 line? i.e. inline dataSourceMetadata calculation